Add EventStore Hosting and Client integrations

fredimachado opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #236**

<!-- Add an overview of the changes here -->
This pull request includes several updates to the Aspire CommunityToolkit including new projects for EventStore Hosting and Client integrations, a simple "bank account" example project, tests and updates to dependencies.

### Project Additions and Updates:
Added new EventStore projects to the solution, including `Aspire.CommunityToolkit.Hosting.EventStore`, `Aspire.CommunityToolkit.EventStore`, and their respective test and example projects.

### Dependency Updates:
Added new dependencies in `Directory.Packages.props`, including `AspNetCore.HealthChecks.EventStore.gRPC`, `EventStore.Client.Extensions.OpenTelemetry`, and `EventStore.Client.Grpc.Streams`. (`Directory.Packages.props`) [[1]](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-5baf5f9e448ad54ab25a091adee0da05d4d228481c9200518fcb1b53a65d4156)

### Code Ownership:
Updated the CODEOWNERS file to include new EventStore projects and assigned ownership to @fredimachado. (CODEOWNERS)

### New Example Project:
Added a simple "bank account" example project for EventStore service. (examples/eventstore/Aspire.CommunityToolkit.Hosting.EventStore.ApiService) [[1]](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-979a3162ee391e64220c83e4cc44fd42a4088f5f4e481a608cbbe7dab5110a04)
<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

## PR Checklist

<!-- Please check if your PR fulfills the following requirements, and remove the ones that are not applicable to the current PR -->

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] New integration
  - [x] Docs are written
  - [x] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

## Other information

Docs PR: https://github.com/dotnet/docs-aspire/pull/2143

aaronpowell commented on CODEOWNERS:
I'll add you as a contributor to address the errors reported here - assuming you are wanting to come onboard as a maintainer.

fredimachado replied:
Sure, I'd love to help :)

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:
```diff
@@ -0,0 +1,210 @@
+    /// </code>
+    /// </example>
+    /// </remarks>
+    public static IResourceBuilder<EventStoreResource> AddEventStore(
+        this IDistributedApplicationBuilder builder, string name, int? port = null)
```
```suggestion
    public static IResourceBuilder<EventStoreResource> AddEventStore(this IDistributedApplicationBuilder builder, [ResourceName] string name, int? port = null)
```

Let's avoid arbitrary line breaks.

Also added the `ResourceName` attribute to the name parameter for the analyzer

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:
```diff
@@ -0,0 +1,210 @@
+        builder.Services.AddHealthChecks()
+            .Add(new HealthCheckRegistration(
+                healthCheckKey,
+                sp => new EventStoreHealthCheck(connectionString!),
```
I'd prefer that we pass the resource into the constructor and then resolve the connection string when the health check is executed.

If the connection string can't be resolved, the resource is unhealthy.

This way we're avoiding relying on the event and we also don't have to use the null forgiving operator.

aaronpowell replied:
Actually, is that a limitation of the external health check package?

fredimachado replied:
Yes,. Unfortunately, it only takes a connection string at the moment.

aaronpowell replied:
All good, I'd forgotten you were using an external library for the health check on initial review (when I wrote the first comment) and then realised at the end.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.EventStore/README.md:
```diff
@@ -0,0 +1,37 @@
+
+### Install the package
+
+In your AppHost project, install the package using the following command:
```
```suggestion
In your app host project, install the package using the following command:
```

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.EventStore/README.md:
```diff
@@ -0,0 +1,37 @@
+# CommunityToolkit.Aspire.Hosting.EventStore library
+
+Provides extension methods and resource definitions for the .NET Aspire AppHost to support running [EventStore](https://www.eventstore.com) containers.
```
```suggestion
Provides extension methods and resource definitions for the .NET Aspire app host to support running [EventStore](https://www.eventstore.com) containers.
```

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.EventStore/README.md:
```diff
@@ -0,0 +1,37 @@
+
+### Example usage
+
+Then, in the _Program.cs_ file of `AppHost`, add a EventStore resource and consume the connection using the following methods:
```
```suggestion
Then, in the _Program.cs_ file of app host, add a EventStore resource and consume the connection using the following methods:
```

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,109 @@
+    {
+        var resourceName = "eventstore";
+        await fixture.ResourceNotificationService
+            .WaitForResourceAsync(resourceName, KnownResourceStates.Running)
```
it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,109 @@
+    {
+        var resourceName = "apiservice";
+        await fixture.ResourceNotificationService
+            .WaitForResourceAsync(resourceName, KnownResourceStates.Running)
```
it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,109 @@
+    {
+        var resourceName = "apiservice";
+        await fixture.ResourceNotificationService
+            .WaitForResourceAsync(resourceName, KnownResourceStates.Running)
```
it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:
```diff
@@ -0,0 +1,109 @@
+    {
+        var resourceName = "apiservice";
+        await fixture.ResourceNotificationService
+            .WaitForResourceAsync(resourceName, KnownResourceStates.Running)
```
it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStorePublicApiTests.cs:
can we also get some tests that the right annotations are added for the volume/data bind mounts

fredimachado replied:
Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

aaronpowell reviewed (changes requested):
I don't see tests for the client integration, only the hosting, I assume that they are still to come

fredimachado commented:
> I don't see tests for the client integration, only the hosting, I assume that they are still to come

Addressed in this commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/8fc728b009144a6e1db5fc155801875268a6d550

aaronpowell reviewed (approved):
Looks really good @fredimachado 

I've kicked off a build, once it's clean we can look to merge it in.

You should also have an invite to the contributor team (which will sort out the CODEOWNERS)

Alirexaa commented on src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:
```diff
@@ -0,0 +1,209 @@
+        ArgumentNullException.ThrowIfNull(builder);
+
+#pragma warning disable CTASPIRE001 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
+        return builder.WithVolume(name ?? VolumeNameGenerator.CreateVolumeName(builder, "log"), LogTargetFolder);
```
Why we have seperate method for this instead of using this in `WithDataVolume`?

fredimachado replied:
This one is for mounting logs (`/var/logs/eventstore`) while the other one mounts data (`/var/lib/eventstore`).
Should we combine both? I usually don't store logs as it can get pretty big.

Alirexaa replied:
If users usually don't store logs that should be fine.

fredimachado commented on src/CommunityToolkit.Aspire.EventStore/AspireEventStoreExtensions.cs:
```diff
@@ -0,0 +1,116 @@
+                sp => new EventStoreHealthCheck(settings.ConnectionString!),
+                failureStatus: default,
+                tags: default,
+                timeout: default));
```
Addressed in this commit https://github.com/CommunityToolkit/Aspire/pull/277/commits/9bb0b5d6e6b3769b21791931f27d077aabc8fef3

Alirexaa commented on tests/CommunityToolkit.Aspire.EventStore.Tests/ConformanceTests.cs:
```diff
@@ -69,6 +69,7 @@ protected override (string json, string error)[] InvalidJsonToErrorMessage => ne
     protected override void SetHealthCheck(EventStoreSettings options, bool enabled)
     {
         options.DisableHealthChecks = !enabled;
+        options.HealthCheckTimeout = TimeSpan.FromMilliseconds(100);
```
```suggestion
        
```

Alirexaa reviewed (approved):
Besides of suggested change totally LGTM.

Alirexaa reviewed (changes requested):
I reviewed tests and did not see tests around persistence and health check functionality.

Could you add these tests?
these are some examples at 
https://github.com/CommunityToolkit/Aspire/blob/main/tests/CommunityToolkit.Aspire.Hosting.Meilisearch/MeilisearchFunctionalTests.cs

fredimachado commented:
> I reviewed tests and did not see tests around persistence and health check functionality.
> 
> Could you add these tests? there are some examples at https://github.com/CommunityToolkit/Aspire/blob/main/tests/CommunityToolkit.Aspire.Hosting.Meilisearch/MeilisearchFunctionalTests.cs

I thought [AppHostTests](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-39c20f8f153ec2acf9f9d92e8ee70f4c7d74c9b6940d22651514c6dbbe67c82a) would also count as persistence tests, since it uses the example API to append and then read events from EventStore.

fredimachado commented:
> > > I reviewed tests and did not see tests around persistence and health check functionality.
> > > Could you add these tests? there are some examples at https://github.com/CommunityToolkit/Aspire/blob/main/tests/CommunityToolkit.Aspire.Hosting.Meilisearch/MeilisearchFunctionalTests.cs
> > 
> > 
> > I thought [AppHostTests](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-39c20f8f153ec2acf9f9d92e8ee70f4c7d74c9b6940d22651514c6dbbe67c82a) would also count as persistence tests, since it uses the example API to append and then read events from EventStore.
> 
> AppHost does not stop across tests. So it's not enough for all cases.

All good. I added the functional tests based on the Meilisearch ones. https://github.com/CommunityToolkit/Aspire/pull/277/commits/d78ff8fa9122e4442452bf77cfbe752df92804b8

aaronpowell reviewed (approved):
I'm happy with how it's looking with the extra tests, but I'll defer to @Alirexaa for final say

fredimachado commented:
@Alirexaa I removed the Log volume/mount since it's probably not going to be used often, and we can always use `WithVolume` extension method in the ResourceBuilder to manually add the log volume if necessary.

fredimachado commented:
@aaronpowell I noticed the SWA example test is stuck during the build's Test step.
I noticed some packages have vulnerabilities, not sure if it has something to do with that. Windows Tests are runnng fine though.
```
Run npm install -g @azure/static-web-apps-cli
(node:3792) ExperimentalWarning: CommonJS module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm warn deprecated rimraf@2.7.1: Rimraf versions prior to v4 are no longer supported

changed 345 packages in 15s

72 packages are looking for funding
  run `npm fund` for details
(node:3822) ExperimentalWarning: CommonJS module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

added 199 packages, and audited [20](https://github.com/CommunityToolkit/Aspire/actions/runs/12020126580/job/33508109977?pr=277#step:18:21)0 packages in 3s

42 packages are looking for funding
  run `npm fund` for details

2 vulnerabilities (1 moderate, 1 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
```

aaronpowell commented:
Yeah, unfortunately that's out of our control, we're installing the latest package version

aaronpowell commented:
It looks like we have a test that is hanging and causing the Linux runner to time out after several hours.

Assuming that these tests pass locally, this is going to be a fun one to try and resolve 🤣. Comparing the test results artifacts between Windows and Linux the only missing TRX file is the `CommunityToolkit.Aspire.Hosting.EventStore.Tests`, which suggests that the problematic tests are in that project.

The most likely candidates for the issue is going to be the tests in the `EventStoreFunctionalTests.cs` file. When I've seen this problem in the past it is related to the test waiting for something to complete that never completes but is still running "successfully" so a timeout on the async method doesn't work.

Here's the steps I'd go through to try and resolve it:

1. Add a `Skip` attribute to all tests in that file and push
  1.1. If the runner still times out, also skip the test that use the example project - basically all tests that use Docker since they don't run on Windows
3. Add one test back per push until you find the test that causes the runner to not exit

Once you know what test is the problematic one, we'll be able to help diagnose the underlying problem.

fredimachado commented:
All good, thanks for the suggestions. I'll try it.

aaronpowell commented:
> All good, thanks for the suggestions. I'll try it.

Unfortunately, this isn't the first time I've seen runners hang 🤣

aaronpowell commented:
Well, looks like we found where the tests are that are the problem

fredimachado commented:
It seems the we had to change permissions for the data bind mount to work on the Ubuntu build.
cc @aaronpowell @Alirexaa

aaronpowell commented:
> It seems the we had to change permissions for the data bind mount to work on the Ubuntu build. cc @aaronpowell @Alirexaa
> 
> <img alt="image" width="778" src="https://private-user-images.githubusercontent.com/29800/390692970-1767ddf1-db36-4d00-8060-bbf8b44625f4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDEzMzksIm5iZiI6MTczMjg0MTAzOSwicGF0aCI6Ii8yOTgwMC8zOTA2OTI5NzAtMTc2N2RkZjEtZGIzNi00ZDAwLTgwNjAtYmJmOGI0NDYyNWY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI5VDAwNDM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlZTZhNDQ5ZDlkZjZlZTc2ZTQ3ODAxN2U1ZTMyM2Y3ZjVkNDYxYzI0MjIyOGZjMWY1MjAxN2U0NTZjNjhhNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tRtSd2TNQdMo-84rNzyBwPrbizZXfIOtRq40O69ogM8">

That's interesting, wonder why...

aaronpowell commented:
Just waiting for another review from @Alirexaa and it should be good to go

Alirexaa commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs:
```diff
@@ -120,16 +131,18 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
             else
             {
+                //EventStore shutdown can be slightly delayed, so second instance might fail to start when using the same bind mount before shutdown.
+                await Task.Delay(TimeSpan.FromSeconds(5));
```
This could incur flakiness.

Related https://github.com/dotnet/aspire/issues/4878

Alirexaa commented on tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs:
```diff
@@ -191,7 +204,7 @@ public async Task VerifyWaitForEventStoreBlocksDependentResources()
                               .WithHealthCheck("blocking_check");
 
-        var dependentResource = builder.AddEventStore("dependentresource")
+        var dependentResource = builder.AddContainer("nginx", "mcr.microsoft.com/cbl-mariner/base/nginx", "1.22")
```
Is this a light (meaning small image size and fast startup) container image?
If it is we could use this image for dependent resources in health check tests like this test to speed up our tests.

fredimachado replied:
Exactly, I used this specifically because it's lightweight and fast to start.
I've seen usages of this image for the same purpose in the main Aspire repo.

Alirexaa reviewed (approved):
@fredimachado Please take a look at my comments.
Otherwise LGTM

Alirexaa commented:
@aaronpowell, CodeQL is blocking me to merge PR.

aaronpowell commented:
> @aaronpowell, CodeQL is blocking me to merge PR.

Grr, I need to find out why that happens

fredimachado commented:
Thanks @aaronpowell and @Alirexaa for your support.
